controllers/storageclass/storageclass_controller.go

```diff
@@ -118,50 +118,71 @@ func (r *StorageClassWatcher) Reconcile(_ context.Context, request reconcile.Req
 		return result, err
 	}
 
-	flashSystemCluster, fscErr := r.getFlashSystemClusterByStorageClass(sc)
-	if fscErr == nil {
-		fscName := flashSystemCluster.GetName()
-		r.Log.Info("FlashSystemCluster found", "FlashSystemCluster", fscName)
-
-		// Check GetDeletionTimestamp to determine if the object is under deletion
-		if !sc.GetDeletionTimestamp().IsZero() {
-			r.Log.Info("object is terminated")
-			err = r.removeStorageClassFromConfigMap(*configMap, fscName, request.Name)
-			if err != nil {
-				return result, err
-			}
-		} else {
-			poolName, ok := sc.Parameters[util.CsiIBMBlockScPool]
-			if ok {
-				if err = r.addStorageClassToConfigMap(*configMap, fscName, request.Name, poolName); err != nil {
+	isTopology := false
+	if _, ok := sc.Parameters["by_management_id"]; ok {
+		isTopology = true
+		r.Log.Info("StorageClass is a topology aware StorageClass", "isTopology", isTopology)
+	}
+	flashSystemClusters, secretMgmtDataByMgmtId, fscErr := r.getFlashSystemClusterByStorageClass(sc, isTopology)
+	for _, fsc := range flashSystemClusters {
+		if fscErr == nil {
+			fscName := fsc.GetName()
+			r.Log.Info("FlashSystemCluster found", "FlashSystemCluster", fscName)
+
+			// Check GetDeletionTimestamp to determine if the object is under deletion
+			if !sc.GetDeletionTimestamp().IsZero() {
+				r.Log.Info("object is terminated")
+				err = r.removeStorageClassFromConfigMap(*configMap, fscName, request.Name)
+				if err != nil {
 					return result, err
 				}
 			} else {
-				r.Log.Error(nil, "cannot reconcile StorageClass without a pool")
+				if isTopology {
+					poolName, err := r.extractPoolName(*sc, secretMgmtDataByMgmtId, fsc)
+					if err != nil {
+						r.Log.Error(err, "failed to extract pool name from topology aware storageClass")
+						return result, err
+					}
+					err = r.addStorageClassToConfigMap(*configMap, fscName, request.Name, poolName)
+					if err != nil {
+						r.Log.Error(err, "failed to add topology aware StorageClass to ConfigMap")
+						return result, err
+					}
+					return result, nil
+				}
+				poolName, ok := sc.Parameters[util.CsiIBMBlockScPool]
+				if ok {
+					if err = r.addStorageClassToConfigMap(*configMap, fscName, request.Name, poolName); err != nil {
+						return result, err
+					}
+				} else {
+					r.Log.Error(nil, "cannot reconcile StorageClass without a pool")
+				}
 			}
+			return result, nil
+		} else {
+			r.Log.Info("cannot find FlashSystemCluster for StorageClass")
+			return result, fscErr
 		}
-		return result, nil
-	} else {
-		r.Log.Info("cannot find FlashSystemCluster for StorageClass")
-		return result, fscErr
 	}
+	return result, nil
 }
 
-func (r *StorageClassWatcher) getFlashSystemClusterByStorageClass(sc *storagev1.StorageClass) (v1alpha1.FlashSystemCluster, error) {
+func (r *StorageClassWatcher) getFlashSystemClusterByStorageClass(sc *storagev1.StorageClass, isTopology bool) ([]v1alpha1.FlashSystemCluster, map[string]interface{}, error) {
```

Comment:
There is something wrong with the logic of this function.
If we have topology StorageClass then we do not need to loop over all FlashSystemClusters, as we have already done this in `getManagementMapFromSecret` function.

Also, I am not sure why do we need to return the `secretMgmtDataByMgmtId` map - is it for extracting the pool name? if so, then I think it's not necessary, as we return from `getManagementMapFromSecret` a mapping between mgmtID and FSC, so we can use the mgmtID with the `by_managment_id` map.